docs/contributing/Powershell Guidelines.md

```diff
@@ -1,123 +1,95 @@
-error. 
+The rationale for these parts are:
 
-If the script is complicated enough that it contains a usage / help display, then the following 
-pattern can be used:
+- `Set-StrictMode`: forces forces Powershell into a more strict mode of interpretation and
+swaps out the "On Error Resume Next" approach for an "On Error Stop" model. Both of these 
+help with our goals of reliability as it makes errors hard stops (unless specifically stated otherwise)
+- `$prepareMachine = $ci`: is necessary to ensure `ExitWithExitCode` properly exits all
+build processes in CI environments (conforming to arcade guidelines).
+- `ExitWithExitCode`: arcade standard for exiting a script that deals with process management
+- `PositionalBinding=$false`: alerts callers when they have incorrect parameters invoking a
+script
 
-``` powershell
-[CmdletBinding(PositionalBinding=$false)]
-param (
-    [switch]$build = $false, 
-    [string]$msbuildDir = "",
-    [parameter(ValueFromRemainingArguments=$true)] $badArgs)
-
-try {
-    if ($badArgs -ne $null) {
-        Print-Usage 
-        exit 1
-    }
-}
+### Coding Guidelines
 
-```
-
-## Executing windows commands
+Use `Exec-*` functions to execute programs or `dotnet` commands. This adds automatic
+error detection on invocation failure, incorrect parameters, etc ... 
 
-Invoking windows commands should be done via the Exec function.  This adds automatic error detection 
-to the invocation and removes the need for error prone if checking after every command.
-
-``` powershell
+```powershell
 # DO NOT
 & msbuild /v:m /m Roslyn.sln
-# DO 
-Exec-Block { & msbuild /v:m /m Roslyn.sln }
-```
+& dotnet build Roslyn.sln
 
-Note this will not work for the rare Windows commands which use 0 as an exit code on failure.  For 
-example robocopy and corflags.
+# DO
+Exec-Command "msbuild" "/v:m /m Roslyn.sln"
+Exec-DotNet "build Roslyn.sln"
+```
 
-In some cases windows commands need to have their argument list built up dynamically.  When that 
-happens do not use Invoke-Expression to execute the command, instead use Exec-Command.  The former
-does not fail when the windows command fails, can invoke powershell argument parsing and doesn't 
-have a mechanism for echoing output to console.  The Exec-Command uses Process directly and can support
-the major functionality needed.
+Scripts that have many executions of `dotnet` commands can store the `dotnet` command in a variable
+and use `Exec-Command` instead.
 
+Call `Test-LastExitCode` after invoking a powershell script to make sure failure is not ignored.
 
-``` powershell
-$command = "C:\Program Files (x86)\Microsoft Visual Studio\Preview\Dogfood\MSBuild\15.0\Bin\MSBuild.exe"
-$args = "/v:m Roslyn.sln"
-if (...) { 
-    $args += " /fl /flp:v=diag"
-}
+```powershell
 # DO NOT
-Invoke-Expression "& $command $args"
+& eng/make-bootstrap.ps1
+Write-Host "Done with Bootstrap"
+
 # DO
-Exec-Command $command $args
+& eng/make-bootstrap.ps1
+Test-LastExitCode
```

Comment:
Why not `Exec-Script`?

Reply:
It is very difficult to pass switch based values between scripts with `Exec-Script` and Powershell. For example the following is tricky: 

```powershell
Exec-Script "eng/build.ps1" "-ci:$ci" 
```

That turns into `-ci:true` which isn't a valid Powershell calling syntax.